File: chanceapi-java/src/main/java/ru/dverkask/chance/ChanceTable.java
package ru.dverkask.chance;

import org.jetbrains.annotations.NotNull;

import java.util.*;

public abstract class ChanceTable<T> implements Iterable<T> {
    protected List<Chance<T>> chances = new ArrayList<>();
    protected ChanceTable() {

    }

    @SafeVarargs public static <T> ChanceTable<T> create(ChanceType type, Chance<T>... chances) {
        return switch (type) {
            case PERCENTAGE -> new PercentageChanceTable<>(chances);
            case WEIGHT -> new WeightChanceTable<>(chances);
        };
    }

    @NotNull @Override public Iterator<T> iterator() {
        return new Iterator<T>() {
            private final Iterator<Chance<T>> innerIterator = chances.iterator();
            @Override public boolean hasNext() {
                return innerIterator.hasNext();
            }

            @Override public T next() {
                return innerIterator.next().item();
            }
        };
    }

    public abstract void addChance(T item, double chance);
    public void addCategory(Category<T> category) {
        for (Chance<T> item : category.getItems()) {
            this.addChance(item.item(), item.chance());
        }
    }

    public abstract T roll();

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("ChanceTable{chances=[");
        for (int i = 0; i < chances.size(); i++) {
            Chance<T> chance = chances.get(i);
            sb.append("Chance{item=").append(chance.item().toString())
                    .append(", chance=").append(chance.chance()).append("}");
            if (i < chances.size() - 1) {
                sb.append(", ");
            }
        }
        sb.append("]}");
        return sb.toString();
    }
}

File: chanceapi-kotlin/src/main/kotlin/ru/dverkask/chance/ChanceTable.kt
package ru.dverkask.chance

abstract class ChanceTable<T> : Iterable<T> {
    protected val chances = mutableListOf<Chance<T>>()

    companion object {
        inline fun <reified T> create(type: ChanceType, vararg chances: Chance<T>): ChanceTable<T> = when (type) {
            ChanceType.PERCENTAGE -> PercentageChanceTable(chances.toList())
            ChanceType.WEIGHT -> WeightChanceTable(chances.toList())
        }
    }

    override fun iterator(): Iterator<T> = object : Iterator<T> {
        private val innerIterator = chances.iterator()
        override fun hasNext(): Boolean = innerIterator.hasNext()
        override fun next(): T = innerIterator.next().item
    }

    abstract fun addChance(item: T, chance: Double)
    fun addCategory(category: Category<T>) {
        category.getItems().forEach { addChance(it.item, it.chance) }
    }

    abstract fun roll(): T?

    override fun toString(): String = buildString {
        append("ChanceTable{chances=[")
        chances.joinTo(this, separator = ", ") { chance ->
            "Chance{item=${chance.item}, chance=${chance.chance}}"
        }
        append("]}")
    }
}